File: src/api/routers/statistics.py
#!/usr/bin/env python3
"""统计信息路由模块.

提供统计数据和违规记录的API端点.
所有接口统一使用领域服务，不提供回退逻辑.
"""
import logging
from datetime import datetime, timedelta
from typing import Any, Dict, List, Optional

from fastapi import APIRouter, HTTPException, Query

router = APIRouter()
logger = logging.getLogger(__name__)

# 领域服务依赖
try:
    from src.services.detection_service_domain import get_detection_service_domain
except ImportError:
    get_detection_service_domain = None


def _ensure_domain_service():
    """确保领域服务可用，如果不可用则抛出HTTP异常."""
    if get_detection_service_domain is None:
        raise HTTPException(status_code=503, detail="检测领域服务不可用，请联系系统管理员")
    service = get_detection_service_domain()
    if service is None:
        raise HTTPException(status_code=503, detail="检测领域服务未初始化，请联系系统管理员")
    return service


@router.get("/statistics/realtime", summary="实时统计接口")
async def get_realtime_statistics() -> Dict[str, Any]:
    """获取实时统计信息.

    Returns:
        实时统计数据，包括当前检测状态、违规统计等.

    Raises:
        HTTPException: 如果领域服务不可用
    """
    try:
        domain_service = _ensure_domain_service()
        result = await domain_service.get_realtime_statistics()
        return result
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"获取实时统计失败: {e}", exc_info=True)
        raise HTTPException(status_code=500, detail=f"获取实时统计失败: {str(e)}")


@router.get("/statistics/detection-realtime", summary="智能检测实时统计接口")
async def get_detection_realtime_statistics() -> Dict[str, Any]:
    """获取智能检测实时统计数据（用于首页检测面板）.

    Returns:
        包含处理效率、FPS、帧数、场景分布、性能监控等数据的字典:
        - processing_efficiency: 处理效率 (0-100)
        - avg_fps: 平均FPS
        - processed_frames: 已处理帧数
        - skipped_frames: 已跳过帧数
        - scene_distribution: 场景分布 {static, dynamic, critical}
        - performance: 性能监控 {cpu_usage, memory_usage, gpu_usage}
        - connection_status: 连接状态 {connected, active_cameras}
        - timestamp: 时间戳

    Raises:
        HTTPException: 如果领域服务不可用
    """
    try:
        domain_service = _ensure_domain_service()
        result = await domain_service.get_detection_realtime_stats()
        return result
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"获取检测实时统计失败: {e}", exc_info=True)
        raise HTTPException(
            status_code=500, detail=f"获取检测实时统计失败: {str(e)}"
        )


@router.get("/statistics/summary", summary="事件统计汇总")
async def get_statistics_summary(
    minutes: int = Query(60, ge=1, le=24 * 60),
    limit: int = Query(1000, ge=1, le=10000),
    camera_id: Optional[str] = Query(None, description="按摄像头过滤（可选）"),
) -> Dict[str, Any]:
    """返回最近 N 分钟内的事件统计与分布.

    可选参数 camera_id 用于仅统计指定摄像头的事件.

    Args:
        minutes: 要查询的最近分钟数.
        limit: 返回样本的最大数量.
        camera_id: (可选) 要筛选的摄像头ID.

    Returns:
        一个包含统计信息的字典.

    Raises:
        HTTPException: 如果领域服务不可用或查询失败
    """
    try:
        domain_service = _ensure_domain_service()
        end_time = datetime.utcnow()
        start_time = end_time - timedelta(minutes=minutes)

        analytics = await domain_service.get_detection_analytics(
            camera_id=camera_id, start_time=start_time, end_time=end_time
        )

        # 处理没有检测记录的情况
        if "detection_statistics" not in analytics:
            # 如果返回的是空记录消息，返回空统计
            return {
                "window_minutes": minutes,
                "total_events": analytics.get("total_records", 0),
                "counts_by_type": {},
                "samples": [],
            }

        stats = analytics.get("detection_statistics", {})

        # 适配原响应结构
        return {
            "window_minutes": minutes,
            "total_events": int(stats.get("total_records", 0)),
            "counts_by_type": stats.get("object_distribution", {}),
            "samples": [],
        }
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"获取统计摘要失败: {e}", exc_info=True)
        raise HTTPException(status_code=500, detail=f"获取统计摘要失败: {str(e)}")


@router.get("/statistics/daily", summary="按天统计事件趋势")
async def get_statistics_daily(
    days: int = Query(7, ge=1, le=90),
    camera_id: Optional[str] = Query(None, description="按摄像头过滤（可选）"),
) -> List[Dict[str, Any]]:
    """返回最近 N 天内的每日事件统计.

    输出：[{date: 'YYYY-MM-DD', total_events: int, counts_by_type: {etype: count}}]

    Args:
        days: 要查询的最近天数.
        camera_id: (可选) 要筛选的摄像头ID.

    Returns:
        一个包含每日统计信息的列表.

    Raises:
        HTTPException: 如果领域服务不可用或查询失败
    """
    try:
        domain_service = _ensure_domain_service()
        result = await domain_service.get_daily_statistics(
            days=days, camera_id=camera_id
        )
        return result
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"获取每日统计失败: {e}", exc_info=True)
        raise HTTPException(status_code=500, detail=f"获取每日统计失败: {str(e)}")


@router.get("/statistics/events", summary="事件列表查询")
async def get_statistics_events(
    start_time: Optional[str] = Query(None, description="开始时间 (ISO格式)"),
    end_time: Optional[str] = Query(None, description="结束时间 (ISO格式)"),
    event_type: Optional[str] = Query(None, description="事件类型过滤"),
    camera_id: Optional[str] = Query(None, description="摄像头ID过滤"),
    limit: int = Query(100, ge=1, le=1000, description="返回数量限制"),
) -> Dict[str, Any]:
    """查询事件列表，支持时间范围和多种过滤条件.

    Args:
        start_time: 开始时间 (ISO格式字符串)
        end_time: 结束时间 (ISO格式字符串)
        event_type: 事件类型
        camera_id: 摄像头ID
        limit: 返回数量限制

    Returns:
        包含事件列表的字典

    Raises:
        HTTPException: 如果领域服务不可用或查询失败
    """
    try:
        domain_service = _ensure_domain_service()

        # 解析时间范围
        from datetime import datetime as dt

        start_dt = None
        end_dt = None

        if start_time:
            try:
                start_dt = dt.fromisoformat(start_time.replace("Z", "+00:00"))
            except Exception as e:
                raise HTTPException(status_code=400, detail=f"开始时间格式错误: {str(e)}")

        if end_time:
            try:
                end_dt = dt.fromisoformat(end_time.replace("Z", "+00:00"))
            except Exception as e:
                raise HTTPException(status_code=400, detail=f"结束时间格式错误: {str(e)}")

        result = await domain_service.get_event_history(
            start_time=start_dt,
            end_time=end_dt,
            event_type=event_type,
            camera_id=camera_id,
            limit=limit,
        )
        return result
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"获取事件列表失败: {e}", exc_info=True)
        raise HTTPException(status_code=500, detail=f"获取事件列表失败: {str(e)}")


@router.get("/statistics/history", summary="近期事件历史")
async def get_statistics_history(
    minutes: int = Query(60, ge=1, le=24 * 60),
    limit: int = Query(100, ge=1, le=1000),
    camera_id: Optional[str] = Query(None, description="按摄像头过滤（可选）"),
) -> List[Dict[str, Any]]:
    """返回近期事件列表，按时间倒序.

    Args:
        minutes: 要查询的最近分钟数.
        limit: 返回事件的最大数量.
        camera_id: (可选) 要筛选的摄像头ID.

    Returns:
        一个包含事件详细信息的列表.

    Raises:
        HTTPException: 如果领域服务不可用或查询失败
    """
    try:
        domain_service = _ensure_domain_service()
        result = await domain_service.get_recent_history(
            minutes=minutes, limit=limit, camera_id=camera_id
        )

        # 转换为API响应格式
        formatted_result = []
        for event in result:
            formatted_result.append(
                {
                    "ts": event.get("timestamp"),
                    "camera_id": event.get("camera_id"),
                    "type": event.get("type"),
                    "track_id": event.get("track_id"),
                    "region": event.get("region"),
                    "detail": event.get("metadata", {}),
                }
            )
        return formatted_result
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"获取近期历史失败: {e}", exc_info=True)
        raise HTTPException(status_code=500, detail=f"获取近期历史失败: {str(e)}")

File: frontend/src/api/modules/statistics.ts
/**
 * 统计分析 API
 * 适配后端实际接口
 */

import request from '../index'

// ===== 后端实际返回的数据结构 =====

export interface RealtimeStatisticsResponse {
    timestamp: string
    system_status: string
    detection_stats: {
        total_detections_today: number
        handwashing_detections: number
        disinfection_detections: number
        hairnet_detections: number
        violation_count: number
    }
    region_stats: {
        active_regions: number
        monitored_areas: any[]
    }
    performance_metrics: {
        average_processing_time: number
        detection_accuracy: number
        system_uptime: string
    }
    alerts: {
        active_alerts: number
        recent_violations: any[]
    }
}

export interface DetectionRealtimeResponse {
    processing_efficiency: number
    avg_fps: number
    processed_frames: number
    skipped_frames: number
    scene_distribution: {
        static: number
        dynamic: number
        critical: number
    }
    performance: {
        cpu_usage: number
        memory_usage: number
        gpu_usage: number
    }
    connection_status: {
        connected: boolean
        active_cameras: number
    }
    timestamp: string
}

export interface DailyStatistics {
    date: string
    total_events: number
    counts_by_type: Record<string, number>
}

// ===== 前端使用的数据结构 =====

export interface SystemStatus {
    status: 'online' | 'warning' | 'offline'
    onlineCameras: number
    totalCameras: number
    realtimeFps: number
    activeAlerts: number
}

export interface DetectionTrend {
    date: string
    detections: number
    violations: number
}

export interface ViolationType {
    type: string
    count: number
    percentage: number
}

// ===== API 方法 =====

/**
 * 获取实时统计数据（原始）
 */
export const getRealtimeStatistics = (): Promise<RealtimeStatisticsResponse> => {
    return request.get('/api/v1/statistics/realtime')
}

/**
 * 获取检测实时统计（原始）
 */
export const getDetectionRealtimeStatistics = (): Promise<DetectionRealtimeResponse> => {
    return request.get('/api/v1/statistics/detection-realtime')
}

/**
 * 获取每日统计数据
 */
export const getDailyStatistics = (params: {
    days?: number
    camera_id?: string
}): Promise<DailyStatistics[]> => {
    return request.get('/api/v1/statistics/daily', { params })
}

/**
 * 获取统计摘要
 */
export const getStatisticsSummary = (params: {
    minutes?: number
    limit?: number
    camera_id?: string
}): Promise<{
    window_minutes: number
    total_events: number
    counts_by_type: Record<string, number>
    samples: any[]
}> => {
    return request.get('/api/v1/statistics/summary', { params })
}

/**
 * 获取事件历史
 */
export const getStatisticsHistory = (params: {
    minutes?: number
    limit?: number
    camera_id?: string
}): Promise<any[]> => {
    return request.get('/api/v1/statistics/history', { params })
}

// ===== 组合方法（前端使用） =====

/**
 * 获取系统状态（组合多个接口）
 * 用于首页状态栏
 */
export const getSystemStatus = async (): Promise<SystemStatus> => {
    try {
        const [realtimeStats, detectionStats] = await Promise.all([
            getRealtimeStatistics(),
            getDetectionRealtimeStatistics()
        ])

        return {
            status: realtimeStats.system_status === 'active' ? 'online' : 'offline',
            onlineCameras: detectionStats.connection_status.active_cameras,
            totalCameras: detectionStats.connection_status.active_cameras, // 暂时相同
            realtimeFps: Math.round(detectionStats.avg_fps),
            activeAlerts: realtimeStats.alerts.active_alerts
        }
    } catch (error) {
        console.error('获取系统状态失败:', error)
        throw error
    }
}

/**
 * 获取检测趋势（7天）
 */
export const getDetectionTrend = async (days: number = 7): Promise<DetectionTrend[]> => {
    try {
        const dailyStats = await getDailyStatistics({ days })

        return dailyStats.map(day => {
            // 计算总违规数（所有类型的违规数之和）
            const violations = Object.values(day.counts_by_type).reduce((sum, count) => sum + count, 0)

            return {
                date: day.date,
                detections: day.total_events,
                violations
            }
        })
    } catch (error) {
        console.error('获取检测趋势失败:', error)
        throw error
    }
}

/**
 * 获取违规类型分布
 * 注意：此接口后端暂未实现，使用 daily 数据聚合
 */
export const getViolationTypes = async (days: number = 7): Promise<ViolationType[]> => {
    try {
        const dailyStats = await getDailyStatistics({ days })

        // 聚合所有天的违规类型
        const typeMap: Record<string, number> = {}
        let totalViolations = 0

        dailyStats.forEach(day => {
            Object.entries(day.counts_by_type).forEach(([type, count]) => {
                typeMap[type] = (typeMap[type] || 0) + count
                totalViolations += count
            })
        })

        // 转换为数组并计算百分比
        return Object.entries(typeMap).map(([type, count]) => ({
            type,
            count,
            percentage: totalViolations > 0 ? Number(((count / totalViolations) * 100).toFixed(1)) : 0
        })).sort((a, b) => b.count - a.count) // 按数量降序
    } catch (error) {
        console.error('获取违规类型分布失败:', error)
        throw error
    }
}
